### PR TITLE
CDF forecasts spec

### LINKED ISSUE #63: value type to interval value type
As in https://github.com/SolarArbiter/SolarArbiter.github.io/pull/62

### LINKED ISSUE #64: fix test coverage
this `coveralls.io.yml` line needs to be changed or the `test` directory needs to be renamed to `tests`: https://github.com/SolarArbiter/solarforecastarbiter-api/blob/master/.codecov.yml#L27

I don't care which way we go. We're using `tests` in the core repo.

### PR DESCRIPTION
**Pull request by lboeman**
PR Summary:
- Added the Spec for CDF Forecast endpoints. I added them to the forecasts.py file, because they will live in the forecast blueprint.
- Moved forecast endpoints to forecasts/single and forecasts/cdf respectively to differentiate between single-valued forecasts and probabilistic.
- Added descriptions for '_links' fields that were previously empty in the API spec because APISPEC module and redoc aren't rendering the contents of the Hyperlinks marshmallow fields. I'll have to add descriptions to sites and observations.
Also closes #64, closes #63

### CODE REVIEW

**Comment by lboeman** [comment]
Added the Spec for CDF Forecast endpoints. I added them to the forecasts.py file, because they will live in the forecast blueprint. 

Added descriptions for '_links' fields that were previously empty in the API spec because APISPEC module and redoc aren't rendering the contents of the Hyperlinks marshmallow fields.

File: sfa_api/forecasts.py [hunk 1]
```diff
@@ -328,6 +326,186 @@ def get(self, forecast_id, *args):
...
+            $ref: '#/components/responses/404-NotFound'
+        """
+        return jsonify({'errors': {'error': ['Not Implemented']}})
+
+
+class CDFForecastMetadata(MethodView):
+    def get(self, forecast_id):
+        """
+        ---
+        summary: Get Probabilistic Forecast value.
```
Comment (alorenzo175) [inline sfa_api/forecasts.py]: I would make this 'Get Metadata for one Probabilistic Forecast constant value' or something else more descriptive

File: sfa_api/forecasts.py [hunk 2]
```diff
@@ -328,6 +326,186 @@ def get(self, forecast_id, *args):
...
+            $ref: '#/components/responses/404-NotFound'
+        """
+        return jsonify({'errors': {'error': ['Not Implemented']}})
+
+
+class CDFForecastValues(MethodView):
+    def get(self, forecast_id):
+        """
+        ---
+        summary: Get Probabilistic Forecast data.
```
Comment (alorenzo175) [inline sfa_api/forecasts.py]: 'Get Probabilistic Forecast data for one constant value'

File: sfa_api/forecasts.py [hunk 3]
```diff
@@ -328,6 +326,186 @@ def get(self, forecast_id, *args):
...
+            $ref: '#/components/responses/401-Unauthorized'
+          404:
+            $ref: '#/components/responses/404-NotFound'
+        """
+        return jsonify({'errors': {'error': ['Not Implemented']}})
+
+    def post(self, forecast_id):
+        """
+        ---
+        summary: Add Probabilistic Forecast data
```
Comment (alorenzo175) [inline sfa_api/forecasts.py]: ... for one constant value

File: sfa_api/schema.py [hunk 4]
```diff
@@ -10,10 +10,9 @@
              'poa', 'ac_power', 'dc_power', 'pdf_probability',
              'cdf_value']
```
Comment (alorenzo175) [inline sfa_api/schema.py]: I think with the CDFforecast defined we can get rid of pdf_probability and cdf_value

File: sfa_api/schema.py [hunk 5]
```diff
@@ -329,9 +351,86 @@ class Meta:
...
+            'values': ma.AbsoluteURLFor('forecasts.values',
+                                        forecast_id='<forecast_id>')
+        },
+        description="Contains links to the values and metadata endpoints."
+    )
+
+
+@spec.define_schema('CDFForecastGroupDefinition')
+class CDFForecastGroupPostSchema(ForecastPostSchema):
+    axis = ma.String(
```
Comment (alorenzo175) [inline sfa_api/schema.py]: maybe this isn't how it's done  now, but can these be defined elsewhere (maybe just in to a variable outside a class) so you don't have to repeat

File: sfa_api/schema.py [hunk 6]
```diff
@@ -329,9 +351,86 @@ class Meta:
...
+        description=('The variable values or percentiles for the set of '
+                     'forecasts in the probabilistic forecast.'),
+    )
+
+
+@spec.define_schema('CDFForecastMetadata')
+class CDFForecastSchema(ForecastSchema):
+    _links = ma.Hyperlinks(
+        {
+            'parent': ma.AbsoluteURLFor('forecasts.single_cdf_group',
```
Comment (alorenzo175) [inline sfa_api/schema.py]: I think this should be called group or probabilistic_forecast_group for consistency

File: sfa_api/schema.py [hunk 7]
```diff
@@ -329,9 +351,86 @@ class Meta:
...
+    axis = ma.String(
+        title='Axis',
+        description=('Axis - The axis on which the constant values of the CDF '
+                     'is specified. The axis can be either x (constant'
+                     'variable values) or y (constant percentiles). The axis '
+                     'is fixed and the same for all forecasts in the '
+                     'probabilistic forecast.'),
+        validate=validate.OneOf(['x', 'y'])
+    )
+    parent = ma.UUID()
```
Comment (alorenzo175) [inline sfa_api/schema.py]: see above

File: sfa_api/schema.py (hunk 6 above)
Comment (lboeman) [reply sfa_api/schema.py]: For the flask view name? e.g. forecasts.probabilistic_forecast _group?

Comment (lboeman) [reply sfa_api/schema.py]: jk you meant the key. got it.

File: sfa_api/schema.py [hunk 8]
```diff
@@ -7,13 +7,11 @@
...
 VARIABLES = ['ghi', 'dni', 'dhi', 'temp_air', 'wind_speed',
-             'poa', 'ac_power', 'dc_power', 'pdf_probability',
-             'cdf_value']
+             'poa', 'ac_power', 'dc_power', 'cdf_value']
 
-OBSERVATION_VALUE_TYPES = ['interval_mean', 'instantaneous']
 
-FORECAST_VALUE_TYPES = ['interval_mean', 'interval_max', 'interval_min',
-                        'interval_median', 'percentile', 'instantaneous']
+VALUE_TYPES = ['interval_mean', 'interval_max', 'interval_min',
```
Comment (wholmgren) [inline sfa_api/schema.py]: `INTERVAL_VALUE_TYPES`?

**Comment by wholmgren** [comment]
Also you guys should figure out how you want to address #64 so that your tests all pass again.

**Comment by alorenzo175** [comment]
lets move test to tests

**Comment by lboeman** [comment]
I moved test to tests, but the check is still failing.

**Pull request merged**